draft PR for refactoring brm function

SermetPekin opened the pull request:
**Draft PR: Refactor `brm()` for clarity and flexibility**
- **Moves the full roxygen documentation** of `brm()` to **`R/brm-doc.R`**, keeping the
  code file lean and making both code and docs easier to navigate.  
- Introduces a lightweight S3 class **`brm_call`** that encapsulates every argument passed to `brm()`.  
- Rewrites the core of **`brm()`** so it can be invoked in two ways:  
  1. the familiar “plain” call (unchanged for existing users), or  
  2. with a pre-constructed `brm_call` object.  
- Each standard `brm()` run now creates its own `brm_call`, attaches it to the resulting `brmsfit`, and uses it internally—making the call history explicit and inspectable.  
- Because a `brm_call` is just a tagged list, it can be modified or reused programmatically before refitting.

```r
# `call_only = T` creates a brm_call
call_1 <- brm(count ~ zAge + zBase * Trt + (1|patient),
            data = epilepsy, family = gaussian() , call_only = T)
# alternatively create_brm_call creates a brm_call without redefining parameters
call_2  <- create_brm_call(count ~ zAge + zBase * Trt + (1|patient),
                          data = epilepsy, family = gaussian())

all.equal(call_1 , call_2)
#TRUE 

fit1 <- brm(call_1)
fit2 <- brm(call_2)

call_2$family <- poisson()
fit2_poisson <- brm(call_2)
```

SermetPekin commented:
- After this change full body of brm method is as below.

```r

#' @rdname brm
#' @export
brm <- function(formula, data= NULL, family = gaussian(), prior = NULL,
                autocor = NULL, data2 = NULL, cov_ranef = NULL,
                sample_prior = "no", sparse = NULL, knots = NULL,
                drop_unused_levels = TRUE, stanvars = NULL, stan_funs = NULL,
                fit = NA, save_pars = getOption("brms.save_pars", NULL),
                save_ranef = NULL, save_mevars = NULL, save_all_pars = NULL,
                init = NULL, inits = NULL, chains = 4,
                iter = getOption("brms.iter", 2000),
                warmup = floor(iter / 2), thin = 1,
                cores = getOption("mc.cores", 1),
                threads = getOption("brms.threads", NULL),
                opencl = getOption("brms.opencl", NULL),
                normalize = getOption("brms.normalize", TRUE),
                control = NULL,
                algorithm = getOption("brms.algorithm", "sampling"),
                backend = getOption("brms.backend", "rstan"),
                future = getOption("future", FALSE), silent = 1,
                seed = NA, save_model = NULL, stan_model_args = list(),
                file = NULL, file_compress = TRUE,
                file_refit = getOption("brms.file_refit", "never"),
                empty = FALSE, rename = TRUE, call_only = FALSE, ...) {

  # if called with a `brm_call` object handle it first
  if(inherits(formula , 'brm_call')){
    args <- formula
    return( .brm_internal(args))
  }

  algorithm <- match.arg(algorithm, algorithm_choices())
  backend <- match.arg(backend, backend_choices())
  normalize <- as_one_logical(normalize)
  silent <- validate_silent(silent)
  iter <- as_one_numeric(iter)
  warmup <- as_one_numeric(warmup)
  thin <- as_one_numeric(thin)
  chains <- as_one_numeric(chains)
  cores <- as_one_numeric(cores)
  init <- use_alias(init, inits)
  threads <- validate_threads(threads)
  opencl <- validate_opencl(opencl)
  future <- as_one_logical(future) && chains > 0L
  seed <- as_one_numeric(seed, allow_na = TRUE)
  empty <- as_one_logical(empty)
  rename <- as_one_logical(rename)
  call_only <- as_one_logical(call_only)
  args <- .brm_collect_args(...)
  class(args) <- c("brm_call" , "list")
  # if call_only is TRUE return a brm_call object
  if(call_only){
    return(args)
  }
  .brm_internal(args)
}

```

SermetPekin commented:
since `file_auto = TRUE` it defines file argument's value using cache (hash value is currently stancode + data+ some brm_call parameters, brms and backend versions)

```r
call_g <- brm(count ~ zAge + zBase * Trt + (1|patient),
              data = epilepsy, family = gaussian(), call_only = T, file_auto = TRUE)
call_g2 <- brm(count ~ zAge + zBase * Trt + (1|patient),
              data = epilepsy, family = gaussian(), call_only = T, file_auto = TRUE)
fit_g <- brm(call_g)
# second call returns cache immediately. 
fit_g2 <- brm(call_g2)
print(fit_g2$brm_call) # or print(call_g)
```

```plaintext
<brm_call>
  Formula : count ~ zAge + zBase * Trt + (1 | patient)
  Data    : data.frame [236 x 9]
  Family  : poisson
  Other arguments (42):
    - prior = <NULL>
    - autocor = <NULL>
    - data2 = <NULL>
    - cov_ranef = <NULL>
    - sample_prior = no
    - sparse = <NULL>
    - knots = <NULL>
    - drop_unused_levels = TRUE
    - stanvars = <NULL>
    - stan_funs = <NULL>
    - fit = NA
    - save_pars = <NULL>
    - save_ranef = <NULL>
    - save_mevars = <NULL>
    - save_all_pars = <NULL>
    - init = <NULL>
    - inits = <NULL>
    - chains = 4
    - iter = 2000
    - warmup = 1000
    - thin = 1
    - cores = 1
    - threads = <list:4>
    - opencl = <list:1>
    - normalize = TRUE
    - control = <NULL>
    - algorithm = sampling
    - backend = cmdstanr
    - future = FALSE
    - silent = 1
    - seed = NA
    - save_model = <NULL>
    - stan_model_args = <list:0>
    - file = cache_brmsfit_292d92a559345298.rds
    - file_compress = TRUE
    - file_refit = on_change
    - file_auto = TRUE
    - empty = FALSE
    - rename = TRUE
    - call_only = TRUE
    - dot_args = <list:0>
    - model_hash = 292d92a559345298
```

SermetPekin commented:
Our .internal_brm() function is structured around the brm_call class.
Thanks to this design, we can easily check, modify, and pass all parameters of a brm() call to other functions with minimal effort.
```r
#' Internal engine to evaluate and fit a *brms* model
#' @noRd
.brm_internal <- function(brm_call) {
  # Check if fit object can be reused from file
  result <- .brm_check(brm_call)
  if (!is.null(result)) {
    return(result)
  }
  # build new or reuse existing fit
  .list <- .build_or_reuse(brm_call)
  if(!.list$needs_refit){
     return(.list$x_from_file) # return x from file
  }
  # empty model returns
  if(brm_call$empty){
    return(.list$x)
  }
  # brmsfit object `x`
  x <- .list$x
  x$brm_call <- brm_call
  # fit happens
  fit_args <- .create_fit_args(brm_call)
  x$fit <- do_call(fit_model, fit_args)
  # rename parameters to have human readable names
  if (brm_call$rename) {
    x <- rename_pars(x)
  }
  if (!is.null(brm_call$file)) {
    x <- write_brmsfit(x, brm_call$file, compress = brm_call$file_compress)
  }
  x
}

```

paul-buerkner commented:
Thank you! There are a lot of changes and I will need to time to check them and edit as needed. Let me know when the PR is ready from your side for review/edits.

SermetPekin commented:
> Thank you! There are a lot of changes and I will need to time to check them and edit as needed. Let me know when the PR is ready from your side for review/edits.

I will refine this PR to focus solely on adding the `brm_call` class and refactoring the `brm()` function- I will be removing hashing functions - , without introducing any new features, to ensure a clean transition. Once this PR is approved and merged, it should be straightforward to add the hashing capability and the file_auto parameter, and to test them, now that we’ll have brm_call in place.

I’ll let you know when it’s ready. Thank you for reviewing this!

SermetPekin commented on R/brm.R:
```diff
@@ -485,124 +492,407 @@ brm <- function(formula, data, family = gaussian(), prior = NULL,
+#'       sampler must be run.}
+#'   }
+#' @noRd
+build_new_brmsfit <- function(call) {
```
I renamed all as brm_call

SermetPekin commented on R/brm.R:
```diff
@@ -485,124 +492,407 @@ brm <- function(formula, data, family = gaussian(), prior = NULL,
+#' @param brm_call A list with class <brm_call>.
+#' @return `brm_call` (invisibly) or an error via `rlang::abort()`.
+#' @noRd
+brm_call_type_check <- function(brm_call) {
```
I have removed this function from the package.

SermetPekin commented:
> I only shortly reviewed the brm.R file, made some edits, mostly to function placement and naming. Generally, please keep the doc with the functions (I reverted you moving the doc). You use a lot of `.name` naming convention which I think you can use more sparsely. I also have some more comments about the current code structure. Please let me know if you have questions. More generally, please stick to the coding style of the package rather than setting up your own. Examples include object and function names as well as type checking and error handling. I want to make sure the whole package continues to read consisently.

I  have refactored it to make it more consistent with the codebase. You are right I am probably using .name convention rather too often. It is just a habit to avoid generic names such as `list`, `name` or `get_*` which probably used everywhere just to get away collisions from the same page. These names are generally just for this page and will not be used outside of this same file. I removed some redundant functions for this PR too. Thank you for the suggestions.

paul-buerkner commented:
I have thought more about it and I don't feel comfortable with the current structure. I support the idea of a call object, which I understand is central to your idea of hasing the brms objects and auto_file. But can we reduce the PR on the aspects central to brm_call? 

By that I mean: Do we really need to split the .brm function body into reuse_existing_brmsfit, build_new_brmsfit  etc.? Yes the function body is relatively long but the logical is clean and easy to follow (at least for me and I need to maintain it). In the process of splitting, you had to create the bit awkward tmp lists which contain all sorts of things to keep track of, which is not a particularily clean in- and output structure. And I would prefer not to have that.

I am sorry about being pedantic about it but I think I wouldn't do my job as developer well if I wasn't.

SermetPekin commented:
> I have thought more about it and I don't feel comfortable with the current structure. I support the idea of a call object, which I understand is central to your idea of hasing the brms objects and auto_file. But can we reduce the PR on the aspects central to brm_call?
> 
> By that I mean: Do we really need to split the .brm function body into reuse_existing_brmsfit, build_new_brmsfit etc.? Yes the function body is relatively long but the logical is clean and easy to follow (at least for me and I need to maintain it). In the process of splitting, you had to create the bit awkward tmp lists which contain all sorts of things to keep track of, which is not a particularily clean in- and output structure. And I would prefer not to have that.
> 
> I am sorry about being pedantic about it but I think I wouldn't do my job as developer well if I wasn't.

Sure. That would also work and probably cleaner. and we can get rid of some of the extra functions. Let me refactor and combine them into .brm and see what we have.

paul-buerkner commented:
2. I would prefer to delete the line, as to not safe the call in the brmsfit object. I agree, once a hash is there, the hash should be stored of course.

paul-buerkner commented:
yes, as long as we don't have a line storing the call inside the brmsfit object, like `x$call <- call` we are fine :-)

SermetPekin commented:
> yes, as long as we don't have a line storing the call inside the brmsfit object, like `x$call <- call` we are fine :-)

:) x$prev_call maybe. In other way we are safe as long as we delete that line. :)

SermetPekin commented:
- As you remind me for backwards compatibility it is safer not to attach any new property to `brmsfit` object as long as we do not need that and since we are just implementing a hash function that will calculate and create a file argument for user in case `file_auto=TRUE` I can make sure not to touch `brmsfit` even to attach a hash value cause we do not need that. We create a `brm_call` object internally it will have a hash property it will create file argument only when (file_auto == TRUE ).  No need a change in the existing structure here.

SermetPekin commented:
I have noticed you have added this line where user gave we have a brmsfit object and it needs to be refit and we override call$backend with x$backend.  
```
call$backend <- x$backend
```
In order to handle some edge cases we can modify default `backend` argument of `brm` to be NULL and can distinguish the situation where user gave a `brmsfit` object and gave a different `backend` explicitly and this time wants to run it with different backend. - Or even user does not know if this which `backend` this fit was run-  We can then handle `backend` argument inside `brm` function and do exactly what user wanted to do in edge cases.
backend = getOption("brms.backend", "rstan"),

paul-buerkner commented:
Yeah, we could but this is just one special case of problem (1) that call and the model actually being fitted do not match in relevant ways when argument fit is specified. So if we hash on the call object but don't consider fit, we will hash the wrong thing. If we solve this problem generally, we will have it solved for backend. No need to change the default input in brm for now.

SermetPekin commented:
> Yeah, we could but this is just one special case of problem (1) that call and the model actually being fitted do not match in relevant ways when argument fit is specified. So if we hash on the call object but don't consider fit, we will hash the wrong thing. If we solve this problem generally, we will have it solved for backend. No need to change the default input in brm for now.

You are right. But in my opinion it is a good idea not to change any property of `brm_call` so that we can check later on.  This is the only property as far as I see that we are changing inside `.brm` function. I propose we can think of `brm_call` as a read only object.

SermetPekin commented:
> Yes, it will. Anyway, I have now changed it back so that backend in call is not overwritten. Before I will eventually merge `brm-hash` into main, this needs to be resolved (more generally; see above). Anyway, it doesn't need to be resolved right now.
> 
> From my side this PR is ready to be merged (into `brm-hash`). The next things (correct hashing, auto-file) you then address in following PRs, right?

Thank you for the review and merge.  I am going to work on that and find a solution for the edge cases while handling hash methods.

SermetPekin commented:
> Great! Thank you for contributing to brms! And thank you for your patience with me as I made comments and edits. I really appreciate it.

Thank you for creating it. I love using brms. Honor to contribute to it if I can.
